pkg/frontend/querymiddleware/retry.go

```diff
@@ -7,21 +7,27 @@ package querymiddleware
 
 import (
 	"context"
+	"errors"
 
 	"github.com/go-kit/log"
 	"github.com/go-kit/log/level"
 	"github.com/prometheus/client_golang/prometheus"
 	"github.com/prometheus/client_golang/prometheus/promauto"
 	"github.com/weaveworks/common/httpgrpc"
 
+	apierror "github.com/grafana/mimir/pkg/api/error"
 	util_log "github.com/grafana/mimir/pkg/util/log"
 )
 
+type intObservation interface {
+	Observe(int)
+}
+
 type retryMiddlewareMetrics struct {
 	retriesCount prometheus.Histogram
 }
 
-func newRetryMiddlewareMetrics(registerer prometheus.Registerer) *retryMiddlewareMetrics {
+func newRetryMiddlewareMetrics(registerer prometheus.Registerer) intObservation {
```

Comment:
I don't see the value in this, I think we could perfectly use `prometheus.Observer` inteface and call `float64` casting in all (1?) places.